janitor/functions.py

```diff
@@ -4055,6 +4264,89 @@ def take_a_if_even(x):
+    comparison applied using the parameters listed
+
+    This method does not change the original DataFrame.
+
+     This function is supposed to emulate SQL and R's
+     implementation of conditional or non-equi joins.
+     This exact implementation though is based on the
+     example found in the request issue, typically a
+     conditional join is based on two tables. We may
+     implement it like that just to keep it closer
+     to the actual definition of the conditional join.
+
+     Notes:
+         - This is an unoptimized implementation
+         - This function treats NaN as 0
+         - join_conditional param only handles
+             <, >, != comparators just like SQL and R's
+             conditional joins / non-equi do
+    ex:
+     company_sales = {
+         'SalesMonth': ['Jan', 'Feb', 'Mar', 'April'],
+         'Company1': [150.0, 200.0, 300.0, 400.0],
+         'Company2': [180.0, 250.0, np.nan, 500.0],
+         'Company3': [400.0, 500.0, 600.0, 675.0]
+     }
+     df = pd.DataFrame.from_dict(company_sales)
+
+     The input DataFrame will look as follows
+
+         SalesMonth  Company1  Company2  Company3
+     0        Jan     150.0     180.0     400.0
+     1        Feb     200.0     250.0     500.0
+     2        Mar     300.0       NaN     600.0
+     3      April     400.0     500.0     675.0
+
+     return_df = join_conditional(df, "Company1", "Company2", "<")
+
+     The comparison that is occurring is for each row in the
+     DataFrame, append row if Company 1 value < Company 2 value
+
+     The resulting DataFrame will look as follows.
+
+       SalesMonth  Company1  Company2  Company3
+     0        Jan     150.0     180.0     400.0
+     1        Feb     200.0     250.0     500.0
+     2      April     400.0     500.0     675.0
+
+    :param df: A pandas DataFrame.
+    :param left_column: A string that represents the left side of
+         the comparison e.g. 1 < 5; 1 being the left col
+    :param right_column: A string that represents the right side of
+         the comparison e.g. 5 being the right col in above ex
+    :param join_operator: A string that represents the operator
+         in the comparison e.g. < in the above ex.
+    :returns: A pandas DataFrame with rows that matched our comparison.
+
+
+    """
+
+    return_df = pd.DataFrame(columns=df.columns.tolist())
+    curr_in = 0
+    for index, row in df.iterrows():
+        if join_operator == ">":
+            if row[left_column] > row[right_column]:
+                return_df.loc[curr_in] = list(df.loc[index])
+                curr_in += 1
+        elif join_operator == "<":
+            if row[left_column] < row[right_column]:
+                return_df.loc[curr_in] = list(df.loc[index])
+                curr_in += 1
+        elif join_operator == "!=":
+            if row[left_column] != row[right_column]:
+                return_df.loc[curr_in] = list(df.loc[index])
+                curr_in += 1
```

Comment:
This is more of a note to myself than to you, @BaritoneBeard, but I think there is a cleaner way to handle this chunk of code without the if/elif/else pattern and code duplication that I see here. While this pattern is much more explicit, it also involves more duplication, which makes extending the function with other operators a bit more challenging.